File: 2018/01/rs/src/main.rs
use std::collections::HashSet;

fn part2(changes: &Vec<i32>) -> i32 {
    let changes_length = changes.len();
    let mut frequency = 0;
    let mut previous = HashSet::new();
    let mut index = 0;
    while previous.insert(frequency) {
        frequency += changes[index];
        index = (index + 1) % changes_length;
    };
    frequency
}

fn solve(changes: &Vec<i32>) -> (i32, i32) {
    (changes.iter().sum(), part2(changes))
}

fn get_input(file_path: &String) -> Vec<i32> {
    std::fs::read_to_string(file_path).expect("Error reading input file!").split("\n")
        .map(|line| line.parse().expect(&format!("Unable to parse '{}'", line))).collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P2: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2018/01/cpp/main.cpp
#include <iostream>
#include <fstream>
#include <ctime>
#include <vector>
#include <numeric>
#include <unordered_set>

using namespace std;

struct Results
{
    int part1;
    int part2;
};

int part2(vector<int> changes)
{
    int changes_length = changes.size();
    int frequency = 0;
    unordered_set<int> previous;
    int index = 0;
    while (previous.find(frequency) == previous.end())
    {
        previous.insert(frequency);
        frequency += changes[index];
        index = (index + 1) % changes_length;
    }
    return frequency;
}

Results solve(vector<int> changes)
{
    int sum;
    return {
        accumulate(changes.begin(),
                   changes.end(), 0),
        part2(changes)};
}

vector<int> getInput(char *filePath)
{
    ifstream file(filePath);
    if (!file.is_open())
        throw runtime_error("Error reading input file!");
    string line;
    vector<int> changes;
    while (getline(file, line))
        changes.push_back(stoi(line));
    file.close();
    return changes;
}

int main(int argc, char *argv[])
{
    if (argc != 2)
        throw runtime_error("Please, add input file path as parameter");

    clock_t begin = clock();
    auto results = solve(getInput(argv[1]));
    clock_t end = clock();
    auto elapsed_secs = double(end - begin) / CLOCKS_PER_SEC;
    cout << "P1: " << results.part1 << endl;
    cout << "P2: " << results.part2 << endl;
    cout << endl;
    cout.precision(7);
    cout << "Time: " << std::fixed << elapsed_secs << endl;
    return 0;
}